symengine/symengine_rcp.h

```diff
@@ -198,6 +198,73 @@ using Teuchos::print_stack_on_segfault;
 #endif
 
 
+template <class T>
+class EnableRCPFromThis {
+// Public interface
+public:
+    //! Get RCP<T> pointer to self (it will cast the pointer to T)
+    inline RCP<T> rcp_from_this() {
+#if defined(WITH_SYMENGINE_RCP)
+        return rcp(static_cast<T*>(this));
+#else
+        return rcp_static_cast<T>(weak_self_ptr_.create_strong());
+#endif
+    }
+
+    //! Get RCP<const T> pointer to self (it will cast the pointer to const T)
+    inline RCP<const T> rcp_from_this() const {
```

Comment:
The standard is doing the right thing here.  The reason you need two overloaded functions is because the underlying weak pointer that is stored is of type `weak_ptr<T>`.  You need to function:
```
shared_ptr<T const> shared_from_this() const;
```
so that it will return a `shared_ptr<T const>` so it will not give up a non-const shared ptr to the object.  Otherwise, because C++ only enforces bitwise const, the function:
```
shared_ptr<T> shared_from_this() const
{  return weak_self_ptr_.lock(); }
```
would compile just fine and it would allow a client to change the object  For example, this would allow terrible code like:
```
void someFunc(const SomeType &obj)
{
  obj->shared_from_this()->modify(...);
}
```
where `SomeType::modify()` is a non-const member that changes the state of the object!

When programing in C++ and making things const correct, you have to implement "logical const" not just bitwise const.  The C++ compiler is too dumb to understand logical const.  We as programmers have to do that.  But if we design our classes and functions correctly, the compiler will enforce const for us (and only a const_cast<>() or a C-style cast can remove const and allow modification).  I love this about C++.